Govrn-HQ/govrn-monorepo
# Add guild.xyz integration

## Linear Ticket
[PRO-692 - Guild Integration Consumer](https://linear.app/govrn/issue/PRO-692/guild-integration-consumer)

## Description
Main Changes:
1. Integrated NATS Messaging to establish a connection and retrieve messages containing the guild ID from #358 with couple of changes. _visit `helpers/nats.ts` for more documentations_.
2. Made API calls to fetch guild's basic info, roles and users associated with these roles using [@guildxyz/sdk](https://www.npmjs.com/package/@guildxyz/sdk).
3. Added the required API calls to store and log these fetched data. You can review the permissions file to check the newly added APIs.  _`helpers/db.ts` file for used calls_.
4. There are no changes to the database schema.

### Caveat
In order to a guild import history record to the database, guild id is needed. Consequenlty, If failed to fetch any data from guild's sdk, there will be no record for a failed guild import to show in import history table.

### API examples
- https://api.guild.xyz/v1/guild/caribbean-blockchain-alliance
- https://api.guild.xyz/v1/guild/our-guild (for larger set of data: 60k members)
- https://api.guild.xyz/v1/role/25305

### Required .env values
```
CHAIN_TYPE_ID=1
PROTOCOL_URL="http://localhost:4000/graphql"
GUILD_IMPORT_TOKEN=<your token>
```


### Testing & Debugging
To fetch a guild named `our-guild` from the NATS server stream named `guild-import-job`, follow these steps:
1. Install `nats-server` locally [via a Package Manager](https://docs.nats.io/running-a-nats-service/introduction/installation#installing-via-a-package-manager).
2. Install [NATS Command Line Interface](https://github.com/nats-io/natscli). _This step is optional but highly recommended for a smoother debugging experience_.
3. Add env variables mentioned in the previous section. 
6. Run protocol API: `yarn nx run protocol-api:serve`.
7. run nats server: `nats-server --jetstream`
8. Publish messages to nats server with a stream name: `guild-import-job`. You can acheive that using one of two ways: 
   - nats CLI: `nats -s nats://localhost:4222 publish "guild-import-job.row" "our-guild"`. _It is recommended to complete step 7 at least once before running this command_.
   - Alternatively, call the `writeMessages()` utility function from `/helpers/nats.ts` after you successfully establish a connection.
9. To check if a message is successfully published: `nats stream report`. To view messages: `nats stream view`
10. Run guild import job: `yarn nx run guild-import-job:serve`.


Unfortunately, `nats-server` doesn't include a built-in logging or debugging mode. However, nats CLI provide a convenient ways to interact with the server's state using `nats stream help`, accessing various functionalities and insights about nats server.


Here's a Loom screencast to prepare nats-server for testing: 
https://www.loom.com/share/18fad924307f4c14a87fdd0b0ef1458c
